docs/our-approach-to-data.md

```diff
@@ -0,0 +1,267 @@
+
+As an example, consider that there are many variations of a "user" in the application. A user may be the current user, a subscriber to a site, or someone who has left a comment on a story in your Reader feed. Each of these display user data in different ways, and in some cases retrieve the data from different sources. However, they can all be classified as a user, and relations between the user and a display context can be established through references.
+
+The following state tree demonstrates how users, sites, and posts may be interrelated, but where data is normalized in a way such that it is always kept in sync, avoiding duplication, and facilitating lookup.
+
+```json
+{
+	"users": {
+		"items": {
+			"73705554": {
+				"ID": 73705554,
+				"login": "testonesite2014"
+			}
+		}
+	},
+	"sites": {
+		"items": {
+			"2916284": {
+				"ID": 2916284,
+				"name": "WordPress.com Example Blog",
+				"description": "Just another WordPress.com weblog",
+			}
+		},
+		"siteUsers": {
+			"2916284": {
+				"73705554": {
+					"roles": [ "administrator" ]
+				}
+			}
+		}
+	},
+	"posts": {
+		"items": {
+			"34": {
+				"ID": 34,
+				"title": "Hello World!",
+				"site_ID": 2916284,
+				"author_ID": 73705554
+			}
+		}
+	}
+}
+```
+
+### Container Components
+
+First, if you haven't already, you should consider reading the following blog posts, as they help to explain the reasoning behind splitting data and visual concerns:
+
+- [_Container Components_](https://medium.com/@learnreact/container-components-c0e67432e005#.zd590uacw)
+- [_Smart and Dumb Components_](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0#.cwn6alkqw)
+
+With that in mind, we typically have a few concerns when building a component that has data needs:
+
+- Ensuring that the necessary data is available
+- Making the data available to the component
+- Allowing the component to modify the data
+
+The first of these, ensuring that data is available, is one that we'd wish to eliminate. It is unforunate that a developer should concern themselves with the fetching behavior of data, as it would be preferable instead that a component describe its data needs, and that the syncing/fetching behavior be handled behind the scenes automatically. Tools like [Relay](https://facebook.github.io/relay/) get us closer to this reality, though Relay has environment requirements that we cannot currently satisfy. For the time being, we must handle our own data fetching, but we should be conscious of a future in which fetching is not a concern for our components.
+
+Framed this way, we can consider two types of container components: connected components and fetching components.
+
+#### Connected components
```

Comment:
>We have expressed these as `app-components` in devdocs gallery as opposed to pure UI components.

I think of an app component as a specific variant of connected component: one which is specifically tailored to reuse and self-sufficiency. This ties a bit back to my earlier comment at https://github.com/Automattic/wp-calypso/pull/1378#discussion_r51128393 . I anticipate that many (most?) components in a render hierarchy will be connected, so it would be nice to have a clearer distinction between app component and connected component, the latter of which more broadly describes the effect of connecting the component to the state.